File: dinneer/lib/service/notification/notification_service.dart
import 'dart:convert';
import 'dart:io' show Platform;
import 'package:firebase_database/firebase_database.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:http/http.dart' as http;
import '../../config/api_config.dart';

final FlutterLocalNotificationsPlugin _localNotifications =
    FlutterLocalNotificationsPlugin();

const AndroidNotificationChannel _chatChannel = AndroidNotificationChannel(
  'chat_notifications',
  'Mensagens de Chat',
  description: 'Notificações de novas mensagens no chat',
  importance: Importance.high,
);

class NotificationService {
  static final FirebaseMessaging _messaging = FirebaseMessaging.instance;

  static Future<void> initialize(String userId) async {
    // ⚠️ Push notifications are NOT supported on web
    // Skip initialization to avoid service worker errors
    if (kIsWeb) {
      debugPrint('⚠️ Push notifications não são suportadas na web. Pulando inicialização.');
      return;
    }

    try {
      await _messaging.requestPermission(alert: true, badge: true, sound: true);

      // Create the Android notification channel before registering
      await _localNotifications
          .resolvePlatformSpecificImplementation<
              AndroidFlutterLocalNotificationsPlugin>()
          ?.createNotificationChannel(_chatChannel);

      const androidInit = AndroidInitializationSettings('@mipmap/ic_launcher');
      const iosInit = DarwinInitializationSettings();
      await _localNotifications.initialize(
        const InitializationSettings(android: androidInit, iOS: iosInit),
      );

      final token = await _messaging.getToken();
      if (token != null) await _saveToken(userId, token);

      _messaging.onTokenRefresh.listen((newToken) => _saveToken(userId, newToken));

      FirebaseMessaging.onMessage.listen(_handleForegroundMessage);
      
      debugPrint('✅ Notificações push inicializadas com sucesso');
    } catch (e) {
      debugPrint('❌ Erro ao inicializar notificações: $e');
      // Não propaga o erro para não quebrar o login
    }
  }

  static Future<void> _saveToken(String userId, String token) async {
    try {
      await FirebaseDatabase.instance.ref('users/$userId/fcmToken').set(token);
      debugPrint('FCM token salvo para usuário $userId');
    } catch (e) {
      debugPrint('Erro ao salvar FCM token: $e');
    }
  }

  static void _handleForegroundMessage(RemoteMessage message) {
    final notification = message.notification;
    final android = message.notification?.android;

    if (notification == null) return;

    _localNotifications.show(
      notification.hashCode,
      notification.title,
      notification.body,
      NotificationDetails(
        android: AndroidNotificationDetails(
          _chatChannel.id,
          _chatChannel.name,
          channelDescription: _chatChannel.description,
          icon: android?.smallIcon ?? '@mipmap/ic_launcher',
          importance: Importance.high,
          priority: Priority.high,
        ),
        iOS: const DarwinNotificationDetails(),
      ),
      payload: jsonEncode(message.data),
    );
  }

  static Future<void> sendChatNotification({
    required int encontroId,
    required String senderId,
    required String senderName,
    required String messageText,
  }) async {
    try {
      await http.post(
        Uri.parse('${ApiConfig.baseUrl}notification/send-chat'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'id_encontro': encontroId,
          'id_usuario': senderId,
          'nm_usuario': senderName,
          'tx_mensagem': messageText,
        }),
      );
    } catch (e) {
      debugPrint('Notificação falhou (não bloqueante): $e');
    }
  }
}

File: pdm_php/node-backend/src/modules/notification/notification.service.ts
import { BaseService } from '../../database/BaseService';
import { Database } from '../../database/Database';
import { getFirebaseAdmin } from '../../config/firebase';

export class NotificationService extends BaseService {
  constructor(banco: Database) {
    super(banco);
  }

  async sendChatNotification(
    encontroId: number,
    senderId: number,
    senderName: string,
    messageText: string
  ): Promise<void> {
    const result = await this.conexao.query(
      'SELECT id_usuario FROM tb_encontro_usuario_dn WHERE id_encontro = $1',
      [encontroId]
    );

    const recipientIds: number[] = result.rows
      .map((r: { id_usuario: number }) => r.id_usuario)
      .filter((id: number) => id !== senderId);

    if (recipientIds.length === 0) {
      this.banco.setMensagem(0, 'Nenhum destinatário encontrado');
      return;
    }

    const db = getFirebaseAdmin().database();
    const messaging = getFirebaseAdmin().messaging();
    let sent = 0;

    for (const userId of recipientIds) {
      const userIdStr = userId.toString();

      const tokenSnap = await db.ref(`users/${userIdStr}/fcmToken`).get();
      const token: string | null = tokenSnap.val();
      if (!token) continue;

      const countRef = db.ref(`unread_counts/${encontroId}/${userIdStr}`);
      const countSnap = await countRef.get();
      const unreadCount: number = countSnap.val() ?? 0;

      if (unreadCount >= 2) continue;

      const body =
        messageText.length > 100
          ? `${messageText.substring(0, 97)}...`
          : messageText;

      await messaging.send({
        token,
        notification: { title: senderName, body },
        data: { encontroId: encontroId.toString(), type: 'chat_message' },
        android: { priority: 'high' },
        apns: { payload: { aps: { sound: 'default' } } },
      });

      await countRef.set(unreadCount + 1);
      sent++;
    }

    this.banco.setMensagem(0, `${sent} notificação(ões) enviada(s)`);
  }
}
